### DISCUSSION TITLE
Turbo cache doesn't include functions in Vercel

### DISCUSSION THREAD

**Discussion by depsimon** [vercel/turborepo#8528] [Help] [answered]
### Summary

We have a few domains for our Vercel project that point to specific branches of our git repo.

We have:
- app.domain.com -> main (production)
- app-stg.domain.com -> staging (staging)
- app-dev.domain.com -> dev (dev)

On those domains (not in previews), we sometimes stumble upon 404 for the app.

The solution to this is to redeploy and uncheck the "Use project's ignore Build Step" checkbox.

I noticed on the failed deployments that it's always a FULL TURBO deployment and the produced files are only the static assets. All the functions are missing, which is what causes the 404 I guess.

### Additional information

```js
The project is a Remix.run app that lives in a monorepo that has multiple packages and a storybook app.

Any idea how I can make sure those fixed environments are always built without the cache?
Or perhaps I need to adapt the turbo config to make sure to include the functions as well?

Here's my current turbo config at the root of my monorepo.



{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": ["**/.env.*local"],
  "globalEnv": [],
  "tasks": {
    "build": {
      "outputs": ["build/**", ".cache", "storybook-static/**"]
    }
  }
}
```

### Example

Here's the logs for the cached deployment (full turbo)

```sh
@acme/project:build: ✓ built in 5.94s
  Tasks:    1 successful, 1 total
 Cached:    1 cached, 1 total
   Time:    549ms >>> FULL TURBO
Summary:    /vercel/path0/.turbo/runs/2i2zOmyXxMRvbGp2y99uDrcrmSK.json
WARN: The `vercelPreset()` Preset was not detected.
Build Completed in /vercel/output [8s]
Deploying outputs...
Injected preview comments middleware, you can disable this in your project settings.
Deployment completed
Uploading build cache [97.45 MB]...
Build cache uploaded: 1.240s
```

Deployment summary: 305 static assets

Here's the logs for the correct deployment (uncached)

```sh
@acme/project:build: ✓ built in 6.59s
  Tasks:    1 successful, 1 total
 Cached:    0 cached, 1 total
   Time:    27.955s 
Summary:    /vercel/path0/.turbo/runs/2i35Rvip1Xmju80TzjbCXLz0KuT.json
WARN: Failed to resolve dependency "/vercel/path0/node_modules/.pnpm/@sentry+profiling-node@7.116.0/node_modules/@sentry/profiling-node/lib/sentry_cpu_profiler-.node"
WARN: Failed to resolve dependency "@emotion/is-prop-valid"
Build Completed in /vercel/output [50s]
Deploying outputs...
Injected preview comments middleware, you can disable this in your project settings.
Deployment completed
Uploading build cache [97.45 MB]...
Build cache uploaded: 1.118s
```

Deployment summary: 305 static assets, 43 functions

**Comment by anthonyshew** [comment]
It's likely that the directory that the functions are being outputted to aren't being captured by the `outputs` key.

I apologize for not being more familiar with Remix but do you know which directory they are outputted to? Is it different than the static assets? Once you've identified that folder, you should be able to clear this up.

**Reply by depsimon** [reply]
When building a Remix app with Vite, it generates a `./build/client` and a `./build/server` ([docs](https://remix.run/docs/en/main/guides/vite#new-build-output-paths))

I'm not sure how Vercel detects the functions, but I suppose it's something along the lines of extracting the routes from the `./build/server/index.js`.

![screenshot 2457](https://github.com/vercel/turbo/assets/1822289/95cd2f69-4db5-486f-aea7-bd1426732149)

A Remix template that deploys to Vercel seems to use the same outputs as I did: https://github.com/PhilDL/remix-gospel-stack/blob/main/turbo.json

I am not sure how I can investigate further

**Comment by depsimon** [comment, marked as answer]
Update, I was missing the `.vercel/**` in outputs.
